src/portal.cpp

```diff
@@ -0,0 +1,391 @@
+		line->portal_dst = NULL;
+	}
+
+	return (line->portal);
+}
+
+void P_TranslatePortalXY(line_t* src, line_t* dst, fixed_t& x, fixed_t& y)
+{
+	if (!src || !dst)
+		return;
+
+	fixed_t nposx, nposy;	// offsets from line
+
+	// Get the angle between the two linedefs, for rotating
+	// orientation and velocity. Rotate 180 degrees, and flip
+	// the position across the exit linedef, if reversed.
+	angle_t angle =
+			R_PointToAngle2(0, 0, dst->dx, dst->dy) -
+			R_PointToAngle2(0, 0, src->dx, src->dy);
+
+	angle += ANGLE_180;
+
+	// Sine, cosine of angle adjustment
+	fixed_t s = finesine[angle>>ANGLETOFINESHIFT];
+	fixed_t c = finecosine[angle>>ANGLETOFINESHIFT];
+
+	fixed_t tx, ty;
+
+	nposx = x - src->v1->x;
+	nposy = y - src->v1->y;
+
+	// Rotate position along normal to match exit linedef
+	tx = FixedMul(nposx, c) - FixedMul(nposy, s);
+	ty = FixedMul(nposy, c) + FixedMul(nposx, s);
+
+	tx += dst->v2->x;
+	ty += dst->v2->y;
+
+	x = tx;
+	y = ty;
+}
+
+void P_TranslatePortalVXVY(line_t* src, line_t* dst, fixed_t& vx, fixed_t& vy)
+{
+	angle_t angle =
+		R_PointToAngle2(0, 0, dst->dx, dst->dy) -
+		R_PointToAngle2(0, 0, src->dx, src->dy);
+
+	angle += ANGLE_180;
+
+	// Sine, cosine of angle adjustment
+	fixed_t s = finesine[angle>>ANGLETOFINESHIFT];
+	fixed_t c = finecosine[angle>>ANGLETOFINESHIFT];
+
+	fixed_t orig_velx = vx;
+	fixed_t orig_vely = vy;
+	vx = FixedMul(orig_velx, c) - FixedMul(orig_vely, s);
+	vy = FixedMul(orig_vely, c) + FixedMul(orig_velx, s);
+}
+
+void P_TranslatePortalAngle(line_t* src, line_t* dst, angle_t& angle)
+{
+	if (!src || !dst)
+		return;
+
+	// Get the angle between the two linedefs, for rotating
+	// orientation and velocity. Rotate 180 degrees, and flip
+	// the position across the exit linedef, if reversed.
+	angle_t xangle =
+			R_PointToAngle2(0, 0, dst->dx, dst->dy) -
+			R_PointToAngle2(0, 0, src->dx, src->dy);
+
+	xangle += ANGLE_180;
+	angle += xangle;
+}
+
+void P_TranslatePortalZ(line_t* src, line_t* dst, fixed_t& z)
+{
+	// args[2] = 0 - no teleport
+	// args[2] = 1 - adjust by floor difference
+	// args[2] = 2 - adjust by ceiling difference
+
+	if (src->args[2] == 1)
+	{
+		z = z - src->frontsector->floorplane.ZatPoint(src->v1->x, src->v1->y) + dst->frontsector->floorplane.ZatPoint(dst->v2->x, dst->v2->y);
+	}
+	else if (src->args[2] == 2)
+	{
+		z = z - src->frontsector->ceilingplane.ZatPoint(src->v1->x, src->v1->y) + dst->frontsector->ceilingplane.ZatPoint(dst->v2->x, dst->v2->y);
+	}
+}
+
+// calculate shortest distance from a point (x,y) to a linedef
+fixed_t P_PointLineDistance(line_t* line, fixed_t x, fixed_t y)
+{
+	angle_t angle = R_PointToAngle2(0, 0, line->dx, line->dy);
+	angle += ANGLE_180;
+
+	fixed_t dx = line->v1->x - x;
+	fixed_t dy = line->v1->y - y;
+
+	fixed_t s = finesine[angle>>ANGLETOFINESHIFT];
+	fixed_t c = finecosine[angle>>ANGLETOFINESHIFT];
+
+	fixed_t d2x = FixedMul(dx, c) - FixedMul(dy, s);
+
+	return abs(d2x);
+}
+
+void P_NormalizeVXVY(fixed_t& vx, fixed_t& vy)
```

Comment:
There are remains of the non-visual part, but this code isn't currently executed from anywhere, well, at least it shouldn't be.